File: _API_TESTING_GHERKIN_/DEMOAPP003_TYPESCRIPT_PLAYWRIGHT/screenplay/questions/ResponseBody.ts
import type { Question } from "../core/types";
import type { Actor } from "../actors/Actor";
import { LAST_RESPONSE } from "screenplay/support/memory-keys";
import type { APIResponse } from "@playwright/test";

export class ResponseBody implements Question<Record<string, unknown>> {
  static json(): ResponseBody {
    return new ResponseBody();
  }

  async answeredBy(actor: Actor): Promise<Record<string, unknown>> {
    const response = actor.recall<APIResponse>(LAST_RESPONSE);
    if (!response) {
      throw new Error("No response available for body assertion");
    }
    return (await response.json()) as Record<string, unknown>;
  }
}

File: _API_TESTING_GHERKIN_/DEMOAPP004_PYTHON_PLAYWRIGHT/screenplay/questions/response_body.py
"""Question that returns the JSON body of the last response."""

from playwright.sync_api import APIResponse

from screenplay.support.memory_keys import MemoryKeys


class ResponseBody:
    @staticmethod
    def answered_by(actor):
        response: APIResponse = actor.memory.recall(MemoryKeys.LAST_RESPONSE)
        if response is None:
            raise AssertionError("No response stored in memory")
        return response.json()
